File: 09th_Competition/Jongryul/05/2531.py
import sys
input = sys.stdin.readline

n, d, k, c = map(int, input().split())

cnts = [0] * (d+1)
cnt = 0
arr = [int(input()) for _ in range(n)]

for i in range(k):
	idx = arr[i]
	cnts[idx] += 1

for i in cnts:
	if i:
		cnt += 1
mx = 0
if not cnts[c]:
	mx = cnt + 1
else:
	mx = cnt

s, e = 0, k - 1

while s < n - 1:
	cnts[arr[s]] -= 1
	if not cnts[arr[s]]:
		cnt -= 1
	s += 1
	e += 1
	if not cnts[arr[e % n]]:
		cnt += 1
	cnts[arr[e % n]] += 1
	if not cnts[c] and mx < cnt + 1:
		mx = cnt + 1
	if cnt > mx:
		mx = cnt

print(mx)

File: Season 1/09th_Competition/Jongryul/05/2531.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "input.txt";
const input = fs
  .readFileSync(filePath)
  .toString()
  .trim()
  .split("\n")
  .map((line) => line.replace("\r", ""));

const [n, d, k, c] = input[0].split(" ").map(Number);
const arr = [];
const cnts = Array(d + 1).fill(0);
let cnt = 0;
let [s, e] = [0, k - 1];
let mx = 0;

for (let i = 1; i < n + 1; i++) {
  arr.push(Number(input[i]));
}

for (let i = 0; i < k; i++) {
  const idx = arr[i];
  if (!cnts[idx]) {
    cnt++;
  }
  cnts[idx]++;
}

if (!cnts[c]) {
  mx = cnt + 1;
} else {
  mx = cnt;
}

while (s < n) {
  cnts[arr[s]]--;
  if (!cnts[arr[s]]) {
    cnt--;
  }
  s++;
  e++;
  if (!cnts[arr[e % n]]) {
    cnt++;
  }
  cnts[arr[e % n]]++;

  if (mx < cnt) {
    mx = cnt;
  }
  if (!cnts[c] && mx < cnt + 1) {
    mx = cnt + 1;
  }
}

console.log(mx);
